scripts/dts/python-devicetree/src/devicetree/edtlib.py

```diff
@@ -2401,6 +2401,25 @@ def _init_luts(self) -> None:
                             f"node '{node.path}' compatible '{compat}' "
                             f"has unknown vendor prefix '{vendor}'")
 
+        # The raw index into edt.compat2nodes[compat] is used for node
+        # instance numbering within a compatible.
+        #
+        # As a way to satisfy people's intuitions about instance numbers,
+        # we sort this list so enabled instances come first.
+        #
+        # This might look like a hack, but it keeps drivers and
+        # applications which don't use instance numbers carefully working
+        # as expected, since e.g. instance number 0 is always the
+        # singleton instance if there's just one enabled node of a
+        # particular compatible.
+        #
+        # This doesn't violate any devicetree.h API guarantees about
+        # instance ordering, since we make no promises that instance
+        # numbers are stable across builds.
+        for compat, nodes in self.compat2nodes.items():
```

Comment:
@fgrandel why do you believe that this belongs in edtlib instead of gen_defines main? also, if you really think it belongs here, should we take the opportunity to convert it to an insertion sort?

Reply:
Hi @decsny , thanks for reviewing this PR.

> why do you believe that this belongs in edtlib instead of gen_defines main?

Conceptually this property should be read-only, because it is just a materialized view of internal EDT state. IMO it is bad OO practice to change a derived property owned by another object from the outside (breaks encapsulation, introduces path dependency to other usage sites). And the sorting concern seems to fit the general EDT concern well, too.

Do you see any disadvantage in having the same sorting for all clients?

> should we take the opportunity to convert it to an insertion sort?

Yes, why not. As we only have a 0/1 sort, we could change

```
                self.compat2nodes[compat].append(node)
                if node.status == "okay":
                    self.compat2okay[compat].append(node)
```

to something like

```
                if node.status == "okay":
                    self.compat2okay[compat].append(node)
                    self.compat2nodes[compat].insert(0, node)
                else
                    self.compat2nodes[compat].append(node)
```

@benediktibk As it's your PR, probably better if you change that and I rebase later.

Reply:
I agree with @fgrandel about this being an internal detail of the class. Therefore I think the sorting should happen within the class.

Using an insertion sort is not a bad idea. Probably not a big game changer, but it also doesn't hurt.